cmd/ftl/cmd_box.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/TBD54566975/ftl/common/projectconfig"
+	"github.com/TBD54566975/ftl/internal/exec"
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+const boxDockerFile = `FROM {{.BaseImage}}
+
+WORKDIR /root
+
+COPY modules /root
+
+EXPOSE 8891
+EXPOSE 8892
+
+CMD ["/root/ftl", "dev"]
+
+`
+
+type boxCmd struct {
+	Build boxBuildCmd `cmd:"" help:"Build a self-contained Docker container (FTL-in-a-box) for running a set of modules."`
+	Run   boxRunCmd   `cmd:"" help:"Run an FTL-in-a-box container."`
+}
+
+type boxRunCmd struct {
+}
+
+func (b *boxRunCmd) Run() error {
+	return fmt.Errorf("not implemented")
+}
+
+type boxBuildCmd struct {
+	BaseImage   string   `help:"Name of the ftl-box Docker image to use as a base." default:"ftl0/ftl-box:${version}"`
+	Parallelism int      `short:"j" help:"Number of modules to build in parallel." default:"${numcpu}"`
+	Image       string   `arg:"" help:"Name of image to build."`
+	Dirs        []string `arg:"" help:"Base directories containing modules (defaults to modules in project config)." type:"existingdir" optional:""`
+}
+
+func (b *boxBuildCmd) Help() string {
+	return ``
+}
+
+func (b *boxBuildCmd) Run(ctx context.Context, client ftlv1connect.ControllerServiceClient, projConfig projectconfig.Config) error {
+	if len(b.Dirs) == 0 {
+		b.Dirs = projConfig.AbsModuleDirs()
+	}
+	if len(b.Dirs) == 0 {
+		return errors.New("no directories specified")
+	}
+	engine, err := buildengine.New(ctx, client, b.Dirs, buildengine.Parallelism(b.Parallelism))
+	if err != nil {
+		return err
+	}
+	if err := os.Setenv("GOOS", "linux"); err != nil {
+		return fmt.Errorf("failed to set GOOS: %w", err)
+	}
+	if err := os.Setenv("GOARCH", "amd64"); err != nil {
+		return fmt.Errorf("failed to set GOARCH: %w", err)
+	}
+	if err := engine.Build(ctx); err != nil {
+		return fmt.Errorf("build failed: %w", err)
+	}
+
+	workDir, err := os.MkdirTemp("", "ftl-box-")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+	defer os.RemoveAll(workDir) //nolint:errcheck
+	logger := log.FromContext(ctx)
+	logger.Debugf("Copying")
+	if err := engine.Each(func(m buildengine.Module) error {
+		config := m.Config.Abs()
+		destDir := filepath.Join(workDir, "modules", config.Module)
+
+		// Copy deployment artefacts.
+		files, err := buildengine.FindFilesToDeploy(config)
+		if err != nil {
+			return err
+		}
+		files = append(files, filepath.Join(config.Dir, "ftl.toml"))
+		for _, file := range files {
+			relFile, err := filepath.Rel(config.Dir, file)
+			if err != nil {
+				return err
+			}
+			destFile := filepath.Join(destDir, relFile)
+			logger.Debugf(" %s -> %s", file, destFile)
+			if err := copy.Copy(file, destFile); err != nil {
+				return fmt.Errorf("failed to copy %s to %s: %w", file, destFile, err)
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	baseImage := b.BaseImage
+	baseImageParts := strings.Split(baseImage, ":")
+	if len(baseImageParts) == 2 {
```

Comment:
Do we need to error if `len(baseImageParts) != 2`?

Reply:
I dunno why I did this, I just should just use `url.Port()` 🤦